File: lib/bridge_proto/frame.cpp
#include "frame.h"

#include "sysex7.h"

namespace bridge {

// ---- FrameWriter -----------------------------------------------------------

void FrameWriter::put(uint8_t v) {
  if (len_ >= cap_) {
    ok_ = false;
    return;
  }
  buf_[len_++] = v;
}

void FrameWriter::begin(uint8_t cmd) {
  len_ = 0;
  ok_ = true;
  put(kSysExStart);
  put(kManufacturerId);
  put(kMagic);
  put(kProtocolVersion);
  put(cmd);
}

void FrameWriter::u7(uint8_t v) { put(v & 0x7F); }

void FrameWriter::u14(uint16_t v) {
  put(v & 0x7F);
  put(static_cast<uint8_t>((v >> 7) & 0x7F));
}

void FrameWriter::u21(uint32_t v) {
  put(v & 0x7F);
  put(static_cast<uint8_t>((v >> 7) & 0x7F));
  put(static_cast<uint8_t>((v >> 14) & 0x7F));
}

void FrameWriter::u32(uint32_t v) {
  for (int i = 0; i < 5; ++i) put(static_cast<uint8_t>((v >> (7 * i)) & 0x7F));
}

void FrameWriter::bytes(const uint8_t* p, size_t n) {
  for (size_t i = 0; i < n; ++i) {
    if (p[i] & 0x80) {
      ok_ = false;
      return;
    }
    put(p[i]);
  }
}

void FrameWriter::packed(const uint8_t* p, size_t n) {
  if (len_ + sysex7::packedLen(n) > cap_) {
    ok_ = false;
    return;
  }
  len_ += sysex7::pack(p, n, &buf_[len_]);
}

size_t FrameWriter::end() {
  put(kSysExEnd);
  return ok_ ? len_ : 0;
}

// ---- FrameReader -----------------------------------------------------------

FrameReader::Status FrameReader::parse(const uint8_t* msg, size_t n,
                                       FrameReader& out) {
  if (n < 2 || msg[0] != kSysExStart || msg[n - 1] != kSysExEnd)
    return Status::NotSysEx;
  // Header is 4 bytes plus the command byte, then the terminating F7.
  if (n < kHeaderLen + 1) {
    // Too short to even carry our header — but if the first bytes that are
    // present already disagree with us, it is somebody else's message.
    if (n >= 2 && msg[1] != kManufacturerId) return Status::NotOurs;
    if (n >= 3 && msg[2] != kMagic) return Status::NotOurs;
    return Status::TooShort;
  }
  if (msg[1] != kManufacturerId || msg[2] != kMagic) return Status::NotOurs;
  if (msg[3] != kProtocolVersion) return Status::BadVersion;

  out.cmd_ = msg[4];
  out.payload_ = msg;
  out.pos_ = kHeaderLen;
  out.end_ = n - 1;  // exclusive of the trailing F7
  return Status::Ok;
}

bool FrameReader::u7(uint8_t& v) {
  if (remaining() < 1) return false;
  v = payload_[pos_++] & 0x7F;
  return true;
}

bool FrameReader::u14(uint16_t& v) {
  if (remaining() < 2) return false;
  v = static_cast<uint16_t>((payload_[pos_] & 0x7F) |
                            ((payload_[pos_ + 1] & 0x7F) << 7));
  pos_ += 2;
  return true;
}

bool FrameReader::u21(uint32_t& v) {
  if (remaining() < 3) return false;
  v = static_cast<uint32_t>(payload_[pos_] & 0x7F) |
      (static_cast<uint32_t>(payload_[pos_ + 1] & 0x7F) << 7) |
      (static_cast<uint32_t>(payload_[pos_ + 2] & 0x7F) << 14);
  pos_ += 3;
  return true;
}

bool FrameReader::u32(uint32_t& v) {
  if (remaining() < 5) return false;
  v = 0;
  for (int i = 0; i < 5; ++i)
    v |= static_cast<uint32_t>(payload_[pos_ + i] & 0x7F) << (7 * i);
  pos_ += 5;
  return true;
}

bool FrameReader::rest(uint8_t* out, size_t capacity, size_t& len) {
  const size_t n = remaining();
  if (n > capacity) return false;
  for (size_t i = 0; i < n; ++i) out[i] = payload_[pos_ + i];
  pos_ += n;
  len = n;
  return true;
}

bool FrameReader::unpackRest(uint8_t* out, size_t capacity, size_t& len) {
  const auto r = sysex7::unpack(&payload_[pos_], remaining(), out, capacity);
  if (r.status != sysex7::UnpackStatus::Ok) return false;
  pos_ = end_;
  len = r.len;
  return true;
}

}  // namespace bridge

File: host/src/frame.js
// Frame construction and parsing — PROTOCOL.md §1, §3.
// Host-side counterpart of lib/bridge_proto/frame.cpp.

import {
  HEADER_LEN,
  MAGIC,
  MANUFACTURER_ID,
  PROTOCOL_VERSION,
  SYSEX_END,
  SYSEX_START,
} from './constants.js';
import { pack, unpack } from './sysex7.js';

export class FrameBuilder {
  /** @param {number} cmd */
  constructor(cmd) {
    this.bytes = [SYSEX_START, MANUFACTURER_ID, MAGIC, PROTOCOL_VERSION, cmd];
  }

  u7(v) {
    this.bytes.push(v & 0x7f);
    return this;
  }

  u14(v) {
    this.bytes.push(v & 0x7f, (v >>> 7) & 0x7f);
    return this;
  }

  u21(v) {
    this.bytes.push(v & 0x7f, (v >>> 7) & 0x7f, (v >>> 14) & 0x7f);
    return this;
  }

  u32(v) {
    // >>> keeps the shift unsigned; v may legitimately have bit 31 set.
    for (let i = 0; i < 5; i++) this.bytes.push((v / 2 ** (7 * i)) & 0x7f);
    return this;
  }

  /** Appends bytes that are already 7-bit safe. */
  raw(arr) {
    for (const b of arr) {
      if (b & 0x80) throw new RangeError(`0x${b.toString(16)} is not 7-bit safe`);
      this.bytes.push(b);
    }
    return this;
  }

  /** Appends bytes through 7-in-8 packing. */
  packed(arr) {
    for (const b of pack(Uint8Array.from(arr))) this.bytes.push(b);
    return this;
  }

  /** @returns {Uint8Array} the complete SysEx message, F0 … F7 inclusive. */
  build() {
    return Uint8Array.from([...this.bytes, SYSEX_END]);
  }
}

export const ParseStatus = Object.freeze({
  OK: 'ok',
  NOT_SYSEX: 'not-sysex',
  NOT_OURS: 'not-ours',
  BAD_VERSION: 'bad-version',
  TOO_SHORT: 'too-short',
});

/**
 * Validates the header of a complete SysEx message.
 * @param {Uint8Array} msg
 * @returns {{status: string, cmd?: number, cursor?: FrameCursor}}
 */
export function parseFrame(msg) {
  const n = msg.length;
  if (n < 2 || msg[0] !== SYSEX_START || msg[n - 1] !== SYSEX_END) {
    return { status: ParseStatus.NOT_SYSEX };
  }
  if (n < HEADER_LEN + 1) {
    if (n >= 2 && msg[1] !== MANUFACTURER_ID) return { status: ParseStatus.NOT_OURS };
    if (n >= 3 && msg[2] !== MAGIC) return { status: ParseStatus.NOT_OURS };
    return { status: ParseStatus.TOO_SHORT };
  }
  if (msg[1] !== MANUFACTURER_ID || msg[2] !== MAGIC) {
    return { status: ParseStatus.NOT_OURS };
  }
  if (msg[3] !== PROTOCOL_VERSION) return { status: ParseStatus.BAD_VERSION };

  return {
    status: ParseStatus.OK,
    cmd: msg[4],
    cursor: new FrameCursor(msg.subarray(HEADER_LEN, n - 1)),
  };
}

/** Reads typed fields out of a frame payload. Throws on underrun. */
export class FrameCursor {
  /** @param {Uint8Array} payload */
  constructor(payload) {
    this.payload = payload;
    this.pos = 0;
  }

  get remaining() {
    return this.payload.length - this.pos;
  }

  #need(n) {
    if (this.remaining < n) {
      throw new RangeError(`frame underrun: wanted ${n}, have ${this.remaining}`);
    }
  }

  u7() {
    this.#need(1);
    return this.payload[this.pos++] & 0x7f;
  }

  u14() {
    this.#need(2);
    const v = (this.payload[this.pos] & 0x7f) | ((this.payload[this.pos + 1] & 0x7f) << 7);
    this.pos += 2;
    return v;
  }

  u21() {
    this.#need(3);
    const p = this.payload;
    const v =
      (p[this.pos] & 0x7f) |
      ((p[this.pos + 1] & 0x7f) << 7) |
      ((p[this.pos + 2] & 0x7f) << 14);
    this.pos += 3;
    return v;
  }

  u32() {
    this.#need(5);
    let v = 0;
    // Accumulate with multiplication: bit 31 would make | produce a negative.
    for (let i = 0; i < 5; i++) v += (this.payload[this.pos + i] & 0x7f) * 2 ** (7 * i);
    this.pos += 5;
    return v >>> 0;
  }

  /** The rest of the payload, verbatim (still 7-bit). */
  rest() {
    const v = this.payload.subarray(this.pos);
    this.pos = this.payload.length;
    return v;
  }

  /** The rest of the payload, unpacked from 7-in-8. */
  unpackRest() {
    return unpack(this.rest());
  }
}
